### PR TITLE
feat: track whether expressions change length but don't aggregate, and only allow length-changing expressions if they're followed by aggregations in the lazy API

### PR DESCRIPTION
**Pull request by MarcoGorelli**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- Related issue #\<issue number\>
- Closes #\<issue number\>

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added
- [ ] Documented the changes

### CODE REVIEW

File: tests/expr_and_series/unique_test.py [hunk 1]
```diff
@@ -1,20 +1,49 @@
...
+    df = nw.from_native(constructor(data))
+    result = df.select(nw.col("a").unique().sum())
+    expected = {"a": [3]}
     assert_equal_data(result, expected)
 
 
+def test_unique_illegal_combination(constructor: Constructor) -> None:
+    df = nw.from_native(constructor(data))
+    with pytest.raises(LengthChangingExprError):
+        df.select((nw.col("a").unique() + nw.col("b").unique()).sum())
```
Comment (MarcoGorelli) [inline tests/expr_and_series/unique_test.py:44]: an expression like this is so bizarre that I doubt we'd be cutting off many real use-cases...

File: narwhals/_expression_parsing.py [hunk 2]
```diff
@@ -342,3 +343,44 @@ def operation_is_order_dependent(*args: IntoExpr | Any) -> bool:
...
+
+    n_exprs = len([x for x in args if isinstance(x, Expr)])
+    changes_length = any(
+        isinstance(x, Expr) and x._changes_length and not x._aggregates for x in args
+    )
+    if n_exprs > 1 and changes_length:
+        msg = (
+            "Found multiple expressions at least one of which changes length. You can only use one "
+            "length-changing expression at a time, unless it is followed by an aggregation."
+        )
```
Comment (FBruzzesi) [inline narwhals/_expression_parsing.py:375]: I find this message a bit confusing. I imagine you mean that a length changing expression might be used only if it's the only expression.

Maybe:

```suggestion
        msg = (
            "Found multiple expressions at least one of which changes length. Any length-changing expression "
            "can be used in isolation, unless it is followed by an aggregation."
        )
```

File: tests/expr_and_series/unique_test.py (hunk 1 above)
Comment (FBruzzesi) [reply tests/expr_and_series/unique_test.py:44]: Agree! 😂

File: tests/expr_and_series/unique_test.py [hunk 3]
```diff
@@ -1,20 +1,49 @@
...
+
+def test_unique_expr_agg(
+    constructor: Constructor, request: pytest.FixtureRequest
+) -> None:
+    if any(x in str(constructor) for x in ("duckdb", "pyspark")):
+        request.applymarker(pytest.mark.xfail)
+    df = nw.from_native(constructor(data))
+    result = df.select(nw.col("a").unique().sum())
+    expected = {"a": [3]}
     assert_equal_data(result, expected)
```
Comment (FBruzzesi) [inline tests/expr_and_series/unique_test.py:40]: Can we maybe add a test with a length changing and an aggregation together? This should not raise I think.

e.g.

```suggestion


def test_unique_expr_and_agg(
    constructor: Constructor, request: pytest.FixtureRequest
) -> None:
    if any(x in str(constructor) for x in ("duckdb", "pyspark")):
        request.applymarker(pytest.mark.xfail)
    df = nw.from_native(constructor(data))
    result = df.select(nw.col("a").unique().alias("unique"), nw.col("a").sum().alias("sum"))
    expected = {"unique": [1, 2], "sum": [4]}
    
    assert_equal_data(result, expected)


```

WDYT?

**Review by FBruzzesi — COMMENTED** [review]
Thanks @MarcoGorelli ! This was easier to go through than expected 🚀

I left only a couple of comments but I wonder about other edge cases 🤔

File: tests/expr_and_series/unique_test.py (hunk 3 above)
Comment (MarcoGorelli) [reply tests/expr_and_series/unique_test.py:40]: i'm confused, sorry - `expected` has keys of different lengths?

File: narwhals/expr_cat.py [hunk 4]
```diff
@@ -64,4 +64,6 @@ def get_categories(self: Self) -> ExprT:
         return self._expr.__class__(
             lambda plx: self._expr._to_compliant_expr(plx).cat.get_categories(),
             self._expr._is_order_dependent,
+            changes_length=self._expr._changes_length,
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply narwhals/expr_cat.py:67]: great catch, thanks!

File: tests/expr_and_series/unique_test.py (hunk 3 above)
Comment (FBruzzesi) [reply tests/expr_and_series/unique_test.py:40]: Whops, my bad! Just updated the values. It's a length changing and an aggregated that then get broadcasted in the same select context.

Comment (MarcoGorelli) [reply tests/expr_and_series/unique_test.py:40]: thanks! I think this is the kind of operation that sql / pyspark / wouldn't support, and I'm not sure we need to, it could just as well be done with `df.select('a', nw.col('a').sum().alias('sum')).unique('a')`

trying this out for example raises:
```python
In [14]: rel = duckdb.sql('select * from values (1, 4), (1, 2), (2, 3), (2, 4) df(a, b)')

In [15]: duckdb.sql("""
    ...: select unnest(list(distinct a)), sum(b) over () from rel
    ...: """)
---------------------------------------------------------------------------
BinderException                           Traceback (most recent call last)
Cell In[15], line 1
----> 1 duckdb.sql("""
      2 select unnest(list(distinct a)), sum(b) over () from rel
      3 """)

BinderException: Binder Error: column "b" must appear in the GROUP BY clause or must be part of an aggregate function.
Either add it to the GROUP BY list, or use "ANY_VALUE(b)" if the exact value of "b" is not important.
```

It would still be backwards-compatible if we later allowed it - but it's easier to start with a simpler and predictable rule, and if necessary relax later

Comment (FBruzzesi) [reply tests/expr_and_series/unique_test.py:40]: Cool! Resolving then!

**Comment by MarcoGorelli** [comment]
thanks for your review!
